cmd/shell.go

```diff
@@ -273,6 +284,66 @@ func newStatsRegistry(addr string, logger blog.Logger) prometheus.Registerer {
 	return registry
 }
 
+// newOpenTelemetry sets up our OpenTelemtry tracing
+// It returns an object that should be called to shutdown the tracer
+func newOpenTelemetry(serviceName string, config OpenTelemetryConfig) func() {
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(serviceName),
+			semconv.ServiceVersionKey.String(core.GetBuildID()),
+		),
+	)
+	if err != nil {
+		FailOnError(err, "Could not create OpenTelemetry resource")
+	}
+
+	// Use a ParentBased sampler to respect the sample decisions on incoming
+	// traces, and TraceIDRatioBased to randomly sample new traces.
+	sampler := trace.TraceIDRatioBased(config.SampleRatio)
+	if !config.DisableParentSampler {
+		sampler = trace.ParentBased(sampler)
+	}
+
+	opts := []trace.TracerProviderOption{
+		trace.WithResource(r),
+		trace.WithSampler(sampler),
+	}
+
+	if config.StdoutExporter {
+		exporter, err := stdouttrace.New(stdouttrace.WithPrettyPrint())
+		if err != nil {
+			FailOnError(err, "Could not create OpenTelemetry stdout exporter")
+		}
+		opts = append(opts, trace.WithBatcher(exporter))
+	}
+
+	if config.Endpoint != "" {
+		exporter, err := otlptracegrpc.New(context.Background(),
+			otlptracegrpc.WithInsecure(),
+			otlptracegrpc.WithEndpoint(config.Endpoint))
+		if err != nil {
+			FailOnError(err, "Could not create OpenTelemetry OTLP exporter")
+		}
+
+		opts = append(opts, trace.WithBatcher(exporter))
+	}
+
+	tp := trace.NewTracerProvider(opts...)
+	otel.SetTracerProvider(tp)
```

Comment:
nit: prefer not to create a new named variable if it's only going to be used once, esp immediately on the next line
```suggestion
	otel.SetTracerProvider(trace.NewTracerProvider(opts...))
```

Reply:
the tracerProvider is also referenced in the returned func().  I should give it a better name though